internal/cuckoo/cuckoo.go

```diff
@@ -20,71 +20,84 @@ const (
 	HashFunc      = hash.Highway
 )
 
-// Cuckoo represents a 3-way Cuckoo hash table data structure
-// that contains the items, bucket indices of each item and the 3
-// hash functions. The bucket lookup is a lookup table on items which
-// tells us which item should be in the bucket at that index. Upon
-// construction the items slice has an additional nil value prepended
-// so the index of the Cuckoo.items slice is +1 compared to the index
-// of the input slice you use.
-type Cuckoo struct {
-	items        [][]byte
-	hashIndices  []byte
-	bucketLookup []uint64
+// CuckooHasher is the building block of a Cuckoo hash table. It only holds
+// the bucket size and the hashers. This means the CuckooHasher can be used
+// when a full Cuckoo hash table is not needed.
+type CuckooHasher struct {
 	// Total bucket count, len(bucket)
 	bucketSize uint64
 	// 3 hash functions h_0, h_1, h_2
 	hashers [Nhash]hash.Hasher
 }
 
-// NewCuckoo instantiates a Cuckoo struct with a bucket of size Factor * size,
-// a stash and 3 seeded hash functions for the 3-way cuckoo hashing.
-func NewCuckoo(size uint64, seeds [Nhash][]byte) *Cuckoo {
+// NewCuckooHasher instantiates a CuckooHasher struct.
+func NewCuckooHasher(size uint64, seeds [Nhash][]byte) (*CuckooHasher, error) {
 	bSize := max(1, uint64(Factor*float64(size)))
 	var hashers [Nhash]hash.Hasher
+	var err error
 	for i, s := range seeds {
-		hashers[i], _ = hash.New(HashFunc, s)
-	}
-
-	return &Cuckoo{
-		// extra element is "keeper" to which the bucketLookup can be directed
-		// when there is no element present in the bucket.
-		items:        make([][]byte, size+1),
-		hashIndices:  make([]byte, size+1),
-		bucketLookup: make([]uint64, bSize),
-		bucketSize:   bSize,
-		hashers:      hashers,
-	}
-}
-
-// NewDummyCuckoo instantiates a Cuckoo struct that does not allocate buckets.
-func NewDummyCuckoo(size uint64, seeds [Nhash][]byte) *Cuckoo {
-	bSize := max(1, uint64(Factor*float64(size)))
-	var hashers [Nhash]hash.Hasher
-	for i, s := range seeds {
-		hashers[i], _ = hash.New(HashFunc, s)
+		hashers[i], err = hash.New(HashFunc, s)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
```suggestion
		if hashers[i], err := hash.New(HashFunc, s); err != nil {
			return nil, err
		}
```

Reply:
would this not work?

Reply:
This doesn't work because the hashers[i] is already instantiated. That is why I defined the err before. If I define in the function declaration, I need to name the *CuckooHasher which I won't use